### PR TITLE
implement DCLightningModule

### PR DESCRIPTION
**Pull request by Chahalprincy**
# Pull Request Template

## Description

Implements the `DCLightningModule` class for enabling pytorch lightning training.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [x] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.22.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [x] Run `flake8 <modified file> --count` and check no errors
  - [x] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

**Comment by Chahalprincy** [comment]
Hi, I have a few follow up questions related to the PR.

a) Can you look over the locations where I have added the new files and in terms of organization if it looks okay?
b) The fit using the lightning trainer is slow right now because of which the unit test takes a long time to pass, I will work on resolving that.
c) I also haven't done type annotations for mypy yet, will also work on those.

File: deepchem/lightning/dc_lightning_module.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import pytorch_lightning as pl
```
Comment (rbharath) [inline deepchem/lightning/dc_lightning_module.py:1]: We shouldn't make a new top level folder for this feature. Instead, this should go in `deepchem/models/torch_models/lightning.py`

Comment (rbharath) [inline deepchem/lightning/dc_lightning_module.py:1]: Lightning should be a conditional import. See other torch imports for an example

File: deepchem/lightning/dc_lightning_module.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import torch
+
+
+class DCLightningModule(pl.LightningModule):
```
Comment (rbharath) [inline deepchem/lightning/dc_lightning_module.py:5]: Please add a docstring detailing the purpose of the class along with a docstring. See `TorchModel`'s docstring for an example. The docstring should be a couple of paragraphs long and detailed.

File: deepchem/lightning/dc_lightning_module.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+  def __init__(self, dc_model):
+    super().__init__()
+    self.dc_model = dc_model
+
+    self.pt_model = self.dc_model.model
+    self.loss = self.dc_model._loss_fn
+
+  def configure_optimizers(self):
```
Comment (rbharath) [inline deepchem/lightning/dc_lightning_module.py:14]: Please add docstrings to all methods

File: deepchem/lightning/tests/test_dc_lightning_module.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+import deepchem as dc
+from deepchem.models import MultitaskClassifier
+import pytorch_lightning as pl
+import numpy as np
+import torch
+from torch.utils.data import DataLoader
+from deepchem.lightning.dc_lightning_module import DCLightningModule
+
+
+class MolnetDataset(torch.utils.data.Dataset):
```
Comment (rbharath) [inline deepchem/lightning/tests/test_dc_lightning_module.py:11]: This doesn't have anything to do with molnet? Perhaps better to rename to `TestDataset`

File: deepchem/lightning/tests/test_dc_lightning_module.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+
+def collate_dataset_wrapper(batch):
+  return MolnetDatasetBatch(batch)
+
+
+class TestDCLightningModule(unittest.TestCase):
+
+  def test_multitask_classifier(self):
+    tasks, datasets, _ = dc.molnet.load_hiv(featurizer='ECFP',
```
Comment (rbharath) [inline deepchem/lightning/tests/test_dc_lightning_module.py:45]: HIV is a quite large dataset (40K datapoints). Can you replace in the unit test with a small synthetic dataset of ~10 points?

**Comment by peastman** [comment]
>  b) The fit using the lightning trainer is slow right now because of which the unit test takes a long time to pass, I will work on resolving that.

Your MolnetDataset class is adding a lot of overhead, and it's unnecessary.  You can just call `make_pytorch_dataset()` on any DeepChem Dataset, and it will return a PyTorch dataset that's ready to use.  Be sure to specify the `batch_size` argument so it will take care of assembling batches for you.  It will do it in a more efficient way.

File: deepchem/lightning/dc_lightning_module.py (hunk 1 above)
Comment (abhiskk) [reply deepchem/lightning/dc_lightning_module.py]: I think we still need a dedicated folder for this, the new folder is at `deepchem/models/lightning`. Also I think it should not go under `torch_models` as it uses `torch_models` directly.

@rbharath what are your thoughts on this?

Comment (rbharath) [reply deepchem/lightning/dc_lightning_module.py]: I think `deepchem/models/lightning` is fine as a new folder

File: deepchem/models/lightning/dc_lightning_module.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+  import pytorch_lightning as pl  # noqa
+  PYTORCH_LIGHTNING_IMPORT_FAILED = False
+except ImportError:
+  PYTORCH_LIGHTNING_IMPORT_FAILED = True
+
+
+class DCLightningModule(pl.LightningModule):
+  """DeepChem Lightning Module to be used with Lightning trainer.
+
+  Example code
```
Comment (rbharath) [inline deepchem/models/lightning/dc_lightning_module.py:12]: The formatting here and elsewhere in the docstring has issues. Are you planning to fix in #2958 @Chahalprincy? This new class also needs to be added to the docs/ folder to render on readthedocs

Comment (Chahalprincy) [reply deepchem/models/lightning/dc_lightning_module.py:12]: Yes, I will do it in another PR. Sorry, I missed it.

**Pull request merged**